Implement basic game loop

juangutie opened the pull request:
Add basic game loop and simple ASCII Rendering

- Add a Game class which handles the core game logic and add a KeyListener
class to handle keyboard inputs

- Extract the requestAnimationFrame loop from the renderer to the main
function ('src/script.ts')

- Add a game logic step in the requestAnimationLoop which handles simple
up/down/left/right keyboard inputs and renders the result as an ascii
string in the console

juangutie committed:
Add basic game loop and Game, KeyListener classes

Squash of 2 commits:

- Add Game and KeyListener classes

- Add basic game loop and ASCII Rendering:

Extract the requestAnimationFrame loop from the renderer to the main
function ('src/script.ts').

Add a game logic step in the requestAnimationLoop which handles simple
up/down/left/right keyboard inputs and renders the result as an ascii
string in the console.

keenanturley commented on src/renderer/Renderer.ts:
```diff
@@ -1,109 +1,98 @@
+        this.gl = gl;
+        this.scene = scene;
+        this.camera = camera;
+        this.preRender = preRender;
```
We're using the Airbnb ESLint config, which follows their standard. It dictates 2-space tabs, so the tabs in this change will upset our ESLint. This comment applies to the commit in general. It makes the diff very hard to read because every line is changed.

See: https://github.com/airbnb/javascript#whitespace

keenanturley commented on src/script.ts:
```diff
@@ -1,53 +1,85 @@
+    }, 1000);
+    if (!PRINT_FPS) clearInterval(fpsConsoleLog);
+
+    // Begin rendering
+    let animationRequestId = requestAnimationFrame(function step(time: number) {
+        frames++;
+
+        game.step(time);
+        renderer.render(time);
+
+        animationRequestId = requestAnimationFrame(step);
+    });
```
I'd like to get other opinions and research into this decision. My original thought was to allow the Renderer to manage the rendering loop and have the game loop maintain the Scene, changing it as necessary in a game step, which would be reflected in the next frame. 

Our visuals could have animations and would benefit from being in the consistently rendered loop. The game loop is not intended to have a consistent step frequency, as our design is such that game state will only step in response to player input.

However, you could combine these into a single loop and call a function in the Game object that probes for an update, only acting if there is a queued input.

juangutie replied:
That is a good point. Since the game state will only change upon player keyboard/mouse input, I will change the code back to having the requestAnimationFrame in the Renderer and update the game state upon player input instead of checking every frame.

juangutie replied:
With the changes up to 73660e8, the game state only updates when there is a relevant keyboard input (WASD/Arrow Keys). The rendering loop is maintained separate from the game loop.

keenanturley replied:
Cool. To be clear, it's definitely possible doing the single loop would be better. We may run into that realization upon implementing further features. This is my first time having done something like this, so I'm open to arguments that provide more guidance. :)

keenanturley commented on src/game/Game.ts:
```diff
@@ -0,0 +1,128 @@
+}
+
+interface PlayerObject {
+    r: number;
+    c: number;
```
Can we use more descriptive names for these properties? Took me a bit to assume it means `row` and `column`.

keenanturley commented on src/game/Game.ts:
```diff
@@ -0,0 +1,128 @@
+    }
+
+    step(time: DOMHighResTimeStamp) {
+        if (this.keyListener.anyPressed()) {
+            const inputs = this.keyListener.inputs;
+
+            // Attempt to move character
+            const previousR = this.player.r;
+            const previousC = this.player.c;
+
+            if (inputs.moveUp) this.player.r--;
+            else if (inputs.moveDown) this.player.r++;
+            else if (inputs.moveLeft) this.player.c--;
+            else if (inputs.moveRight) this.player.c++;
+
+            if (this.collision(this.player.r, this.player.c)) {
+                this.player.r = previousR;
+                this.player.c = previousC;
+            }
+
+            // Print map
+            console.log(this.textRender(this.map, this.player));
+
+            this.keyListener.clear();
+        }
```
Refactor to use a guard clause here to prevent unnecessary indentation:
```js
step(time: DOMHighResTimeStamp) {
  if (!this.keyListener.anyPressed()) {
    return;
  }
  const inputs = this.keyListener.inputs;

  // Attempt to move character
  const previousR = this.player.r;
  const previousC = this.player.c;
  ...
}
```

keenanturley commented on src/game/Game.ts:
```diff
@@ -0,0 +1,128 @@
+        }
+    }
+
+    textRender(map: Tile[][], player: PlayerObject): string {
```
Do we need to pass the `map` and `player` in here? Can we use `this.map` and `this.player` instead and remove the parameters?

keenanturley commented on src/game/Game.ts:
```diff
@@ -0,0 +1,128 @@
+    }
+
+    textRender(map: Tile[][], player: PlayerObject): string {
+        let textArray: string[][] = map.map((row: Tile[]) => {
```
`textArray` isn't reassigned, make `const`. ESLint should yell at you for this.

keenanturley commented on src/game/Game.ts:
```diff
@@ -0,0 +1,128 @@
+            });
+        });
+
+        if (this.player.r === undefined || this.player.c === undefined) {
```
Guard clause here too:

```js
if (this.player.r && this.player.c) {
  return;
}
for ...
```

keenanturley commented on src/game/Game.ts:
```diff
@@ -0,0 +1,128 @@
+import KeyListener from "./KeyListener";
+
+enum Tile {
+    FLOOR,
+    WALL,
+    UNDEFINED,
+}
```
Not sure if you've seen this commit (ff6490860fac2a59247597c14ffe79d6da7983e0) yet, but as an extension to this `enum Tile`, we might want to consider classes for tiles.

keenanturley commented on src/game/Game.ts:
```diff
@@ -0,0 +1,128 @@
+}
+
+export default class Game {
+    map: Tile[][];
```
A class for a `Map` might be useful too. If we keep track of the entities that exist in a Map we can easily query for the win condition (number of monsters alive == 0). 

If we have separate layers for things (Tiles under Entities) this will help encapsulate things too.

keenanturley reviewed (changes requested):
Can you run this code through ESLint and amend the commits? I have a few requested changes too, along with suggestions for future additions.

I'd like to talk more about the render / game loop coupling decision with the game folks, but I don't think that particular change blocks us yet.

juangutie committed:
Respond to Pull Request #5 Review-810807528

Squash of 7 commits:

- Fix ESLint errors, use guard clauses, & use more descriptive property names

- Add guard clause to loadMapFromText()

- Restore requestAnimationFrame loop to Renderer

- Fix FPS=NaN on first console.log

- Update game state only upon keyboard input event

Move the requestAnimationFrame loop back into the Renderer and call
the game step function only when the KeyListener detects a keyboard
event.

Additionally, reset the fps counter when the module is refreshed
by adding a clearInterval() call in the Renderer's stopRendering()
function

- Update game state only on _relevant_ keyboard input

- Change variable names to be more descriptive

`r` changed to `row`, `c` changed to `column`

keenanturley commented on src/game/Game.ts:
```diff
@@ -113,6 +115,7 @@ export default class Game {
       return;
     }
 
+    // If the string did not contain a player position, find an arbitrary default
```
nit: I think we can be more specific than "arbitrary" here. It places the player on the first available floor tile in the map, starting from the top left.

keenanturley commented on src/game/KeyListener.ts:
```diff
@@ -8,10 +8,11 @@ export default class KeyListener {
     this.keyDownHandler = (key: KeyboardEvent) => {
-      this.registerInput(key);
-      callback();
+      if (this.registerInput(key)) {
+        callback();
+      }
```
Nice fix. I'm thinking of situations that can arise later if we have keys that are bound to actions that don't necessarily advance a turn (pause, open menu). These might be UI tasks though that are input with mouse buttons, so it might be fine. Just something to be aware of.

juangutie replied:
Regarding the FPS counter, I ran into a similar issue in the console.log version. When the renderer is constructed, an interval is created with setInterval() which needs to be removed for hot module replacement by calling clearInterval() in stopRendering() (which is called by module.hot.dispose()). Is it possible to dynamically delete the HTML element in a similar fashion?

keenanturley replied:
Yeah I started making the UI component so I wouldn't have to deal with that. The solution to the interval is to dispose it in the hot reload, but if I have to start doing that for every HTML element that's not scalable.

keenanturley reviewed (approved):
Looks good. Left a couple comments but nothing that needs to be acted on. 

For the FPS counter, I have local changes that will move it to a UI panel outside the canvas so it won't be in the console anymore. I'm running into annoying issues with Parcel hot module reloading and duplicating the HTML element that was dynamically created in JS, so I'm trying to see if I can make it work. If not, I'll disable hot module reloading or migrate to Webpack.

juangutie committed:
Edit comment, variable names, whitepace

Squash of 2 commits:

- Edit comment and variable names

Edit comment to specify exactly how loadMapFromString() chooses a
starting player position if none is included in the string parameter

`previousR` renamed to `previousRow`
`previousC` renamed to `previousColumn`

- Edit whitespace on PLACEHOLDER_MAP

juangutie commented:
To clarify the force-push, I squashed the previous 10 commits into 3 commits. I hope this does not cause too much of a nuisance, and for the future, I will try to find a workflow that does not involve so many micro-commits on the remote branch or require squashing

keenanturley commented:
> To clarify the force-push, I squashed the previous 10 commits into 3 commits. I hope this does not cause too much of a nuisance, and for the future, I will try to find a workflow that does not involve so many micro-commits on the remote branch or require squashing

I think the usual practice is to squash the PR to 1 merge commit that can be referenced with the PR number and individual changes and discussion can be seen from there.
